packages/extension-driver-canner/test/cannerDataSource.spec.ts

```diff
@@ -0,0 +1,316 @@
+import { CannerServer } from './cannerServer';
+import { CannerDataSource, PGOptions } from '../src';
+import { ExportOptions, streamToArray } from '@vulcan-sql/core';
+import { Writable } from 'stream';
+import * as sinon from 'ts-sinon';
+import * as fs from 'fs';
+import { CannerAdapter } from '../src/lib/cannerAdapter';
+
+const pg = new CannerServer();
+let dataSource: CannerDataSource;
+
+// restore all sinon mock/stub before each test
+beforeEach(() => {
+  sinon.default.restore();
+});
+
+it('Data source should be activate without any error when all profiles are valid', async () => {
+  // Arrange
+  dataSource = new CannerDataSource({}, '', [pg.getProfile('profile1')]);
+  // Act, Assert
+  await expect(dataSource.activate()).resolves.not.toThrow();
+});
+
+it('Data source should throw error when activating if any profile is invalid', async () => {
+  // Arrange
+  const profile1 = pg.getProfile('profile1');
+  dataSource = new CannerDataSource({}, '', [
+    profile1,
+    {
+      name: 'wrong-password',
+      type: 'canner',
+      connection: {
+        ...profile1.connection,
+        password: 'wrong-password',
+      } as PGOptions,
+      allow: '*',
+    },
+  ]);
+  // Act, Assert
+  await expect(dataSource.activate()).rejects.toThrow();
+});
+
+// export method should be executed successfully
+it('Data source should export successfully', async () => {
+  fs.mkdirSync('tmp', { recursive: true });
+  dataSource = new CannerDataSource({}, '', [pg.getProfile('profile1')]);
+  await dataSource.activate();
+
+  // Act, Assert
+  await expect(
+    dataSource.export({
+      sql: 'select 1',
+      directory: 'tmp',
+      profileName: 'profile1',
+    } as ExportOptions)
+  ).resolves.not.toThrow();
+  expect(fs.readdirSync('tmp').length).toBe(1);
+
+  // clean up
+  fs.rmSync('tmp', { recursive: true, force: true });
+}, 100000);
+
+it('Data source should throw when fail to export data', async () => {
+  // Arrange
+  sinon.default
+    .stub(CannerAdapter.prototype, 'createAsyncQueryResultUrls')
```

Comment:
Suggest adding the comment to describe what reason so need to use `prototype` to stub the method, not usual solution :)

Reply:
delivered in 8ed2dfd